### PR TITLE
[shared_preferences] Adds allowList to setPrefix method.

### PR DESCRIPTION
**Pull request by tarrinneal**
Creates optional allow list for preference keys when setting custom prefix.

Also makes integration tests more homogenous across platforms. 

Fixes https://github.com/flutter/flutter/issues/128948

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/shared_preferences/shared_preferences_web/lib/shared_preferences_web.dart
Comment (ditman) [inline packages/shared_preferences/shared_preferences_web/lib/shared_preferences_web.dart:73]:
```suggestion
        allowList?.contains(key) ?? true);
```

(Shorter but not sure if more readable :))

Also (and this is probably a worthless micro-optimization): should the "contains" operation happen before the "startsWith", so the where can bail out fast?

**Review by ditman — COMMENTED** [review]
Web bits lgtm!

File: packages/shared_preferences/shared_preferences/example/integration_test/shared_preferences_test.dart
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences/example/integration_test/shared_preferences_test.dart:107]: Do we really want groups to be async?

File: packages/shared_preferences/shared_preferences_platform_interface/lib/shared_preferences_platform_interface.dart
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_platform_interface/lib/shared_preferences_platform_interface.dart:87]: This is a breaking change; you need to make a new method instead (ideally with a parameter object in case we want to add other things later). For maximal backwards compatibility, you could the default only throw if an allowList is provided, and otherwise have it delegate to the existing `getAllWithPrefix`.

https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changing-platform-interface-method-parameters

File: packages/shared_preferences/shared_preferences/lib/shared_preferences.dart
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:41]: We should either be very explicit both here and in the `clear` docs that the allow list doesn't affect `clear`, or we should make it so that it does.

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java:147]: Optional: consider a ternary:

```
final Set<String> allowSet = allowList == null ? null : new HastSet<>(allowList);
```

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/shared_preferences/shared_preferences/example/integration_test/shared_preferences_test.dart
Comment (reidbaker) [inline packages/shared_preferences/shared_preferences/example/integration_test/shared_preferences_test.dart:138]: Consider using constants or final for the multiple usages  of `"stringKey"` and `"boolKey"`. 
Consider naming the variables something that indicates their status. 
```
final String allowedString = 'stringKey';
final String allowedBool = 'boolKey';
final String notAllowedDouble = 'doubleKey';
const Set<String> allowList = <String>{allowedBool, allowedBool};
...
final bool? testBool = prefs.getBool(allowedBool);
expect(testBool, true);

final double? testDouble = prefs.getDouble(notAllowedDouble);
expect(testDouble, null);
```

File: packages/shared_preferences/shared_preferences/lib/shared_preferences.dart
Comment (reidbaker) [reply packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:41]: Not updating clear feels like a foot gun. But I agree that at minimum the documentation for both should be updated.

Comment (reidbaker) [reply packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:41]: I do like that there is a clearWithPrefix.

File: packages/shared_preferences/shared_preferences/lib/shared_preferences.dart
Comment (reidbaker) [inline packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:48]: I am guessing it is overkill but did you consider making this a function that returned a bool instead of a set?

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java
Comment (reidbaker) [reply packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java:147]: Since this is a new parem why not pass in a set instead of a list?

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java
Comment (reidbaker) [inline packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java:149]: This aligns with the documentation but I would expect this function to actually or these values. 
As a developer I want getAllPrefs to Give me everything with my custom prefix and a list of legacy values I used before.

File: packages/shared_preferences/shared_preferences_android/android/src/test/java/io/flutter/plugins/sharedpreferences/SharedPreferencesTest.java
Comment (reidbaker) [inline packages/shared_preferences/shared_preferences_android/android/src/test/java/io/flutter/plugins/sharedpreferences/SharedPreferencesTest.java:95]: consider constant/final here as well.

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java
Comment (reidbaker) [reply packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java:149]: If you go with the OR approach then I think you should also make the new param an optional named parameter.

File: packages/shared_preferences/shared_preferences_android/android/src/test/java/io/flutter/plugins/sharedpreferences/SharedPreferencesTest.java
Comment (reidbaker) [inline packages/shared_preferences/shared_preferences_android/android/src/test/java/io/flutter/plugins/sharedpreferences/SharedPreferencesTest.java:161]: Should the plugin call `clearWithPrefix` also take a list/set of allowed values?

File: packages/shared_preferences/shared_preferences/lib/shared_preferences.dart
Comment (stuartmorgan) [reply packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:48]: We need to do the filtering on the native side, so having it be a Dart function would make thing *way* more complicated.

Comment (reidbaker) [reply packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:48]: ok that makes sense.

File: packages/shared_preferences/shared_preferences/lib/shared_preferences.dart
Comment (tarrinneal) [reply packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:41]: the `clear` method shouldn't be used if the `setPrefix` method has been called. I think an assert that is the case is probably worth adding. Thoughts?

File: packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java
Comment (tarrinneal) [reply packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java:147]: It's coming from a pigeon api, which doesn't support sets.

Comment (tarrinneal) [reply packages/shared_preferences/shared_preferences_android/android/src/main/java/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.java:147]: Though, it could be transformed before being passed into this method.

File: packages/shared_preferences/shared_preferences/example/integration_test/shared_preferences_test.dart
Comment (tarrinneal) [reply packages/shared_preferences/shared_preferences/example/integration_test/shared_preferences_test.dart:107]: That was an unneeded tag anyway. I meant to erase it.

File: packages/shared_preferences/shared_preferences/lib/shared_preferences.dart
Comment (tarrinneal) [reply packages/shared_preferences/shared_preferences/lib/shared_preferences.dart:41]: sorry, I misremembered how that change was handled before. The allowList will be added to clear.

File: packages/shared_preferences/shared_preferences_web/lib/shared_preferences_web.dart
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_web/lib/shared_preferences_web.dart:67]: Let's rename this to the now-more-accurate `_getFilteredKeys` or something like that; we may not be dealing with just Flutter keys now.

File: packages/shared_preferences/shared_preferences_windows/lib/shared_preferences_windows.dart
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_windows/lib/shared_preferences_windows.dart:128]: Optional: This could be written slightly more safely (since it doesn't require a force-unwrap) as `!(key.startsWith(options.prefix) && (options.allowList?.contains(key) ?? true))`.

File: packages/shared_preferences/shared_preferences_linux/lib/shared_preferences_linux.dart
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_linux/lib/shared_preferences_linux.dart:122]: Same note here as for Windows.

**Review by stuartmorgan — COMMENTED** [review]
A couple of small things on the structure (sorry about the churn naming will cause), but all the implementation and tests look great! If you're on board with the changes I suggested feel free to go ahead and split out the platform interface PR without waiting for another review round, otherwise we can discuss the details more here before you change things.

**Comment by tarrinneal** [comment]
@stuartmorgan assuming the tests pass, this should be ready to go now.

File: packages/shared_preferences/shared_preferences/CHANGELOG.md
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences/CHANGELOG.md:1]: I just realized that we should updated [this section](https://github.com/flutter/packages/tree/main/packages/shared_preferences/shared_preferences#migration-and-prefixes) of the README. We should add a second sentence to the third paragraph about this; something like "In this case, you can set an `allowList` that contains only supported preference types."

**Review by stuartmorgan — APPROVED** [review]
LGTM with one README update.

**Pull request merged**
